### PR TITLE
Add TFT-Summoner-v1 support

### PR DESCRIPTION
**Pull request by Crimack**
First attempt at implementing support for part of the TFT API (https://github.com/meraki-analytics/cassiopeia/issues/319).

Started with the TFT summoner API as a way of understanding the code, since it's almost identical to the vanilla summoner API. It feels a little like I've ended up copying the existing summoner wrappers and renaming things, but it's probably just because the two APIs are so similar right now.

Testing:
```
test_access_properties (test.test_tft_summoner.TestTFTSummoner) ... Making call: https://na1.api.riotgames.com/tft/summoner/v1/summoners/by-name/Kalturi
Making call: https://ddragon.leagueoflegends.com/realms/na.json
ok
test_equality (test.test_tft_summoner.TestTFTSummoner) ... Making call: https://na1.api.riotgames.com/tft/summoner/v1/summoners/by-name/Kalturi
Making call: https://na1.api.riotgames.com/tft/summoner/v1/summoners/wlBGzbtpt43sfuhV9m_3WUOB_WuIaD4sZ2dyxF-1ZzlOV1E
ok
```

### CODE REVIEW

File: cassiopeia/core/tft_summoner.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+from .staticdata import ProfileIcon
+from ..dto.tft_summoner import TFTSummonerDto
+
+
+##############
+# Data Types #
+##############
+
+
+class TFTSummonerData(CoreData):
```
Comment (stelar7) [inline cassiopeia/core/tft_summoner.py:20]: TFTSummoner and LeagueSummoner is the same object. The call url is the only difference

Comment (Crimack) [reply cassiopeia/core/tft_summoner.py:20]: Yeah, they're definitely pretty similar (with the exception of the section below `# Special core methods`, which will be different between the two). My thinking here was that while _/lol/summoner/v4_ and _/tft/summoner/v1_ are currently identical, they might evolve independently of each other and return subtly different DTOs in future. The duplication felt incidental for now

Would you prefer to just use Summoner, and put TFT helper methods into there e.g. `Summoner(name='Boatshoes').tft_match_history` and handle unpicking the two if/when changes happen?

Comment (stelar7) [reply cassiopeia/core/tft_summoner.py:20]: What about inheriting from the original summoner object instead of CoreData?

Comment (jjmaldonis) [reply cassiopeia/core/tft_summoner.py:20]: This is a good point. We might side-step it all together by the comment I am about to make below.

**Comment by jjmaldonis** [comment]
First, thanks a ton for the contribution. The PR looks great. I don't see anything that needs to be changed to merge it.

That said, deciding whether to merge this PR opens up a totally different can of worms, completely independent of the PR itself. It's related to "What is Cassiopeia meant to be?". When this discussion was originally brought up in #319, I delayed this conversation, but now we've got to have it.

So, what is Cassiopeia meant to be? (and what is it now?) Cass was written when only the League of Legends API existed. It was meant to be "a representation of all League of Legends data" that we have access to. This was a grand goal, and it never completely came to pass, but we did pretty dang good. One thing we're quite happy with is how straightforward it is to integrate Cass with other League functionality, such as the role identification codes (both Meraki's and Canisback's role ID code work with Cass). Cass is also incredibly flexible when it comes to working with other apps, and while that can take a lot of work, it's really really powerful. We're definitely proud of this.

Since Cass was first written, the idea of "all League of Legends data" has changed significantly. We now have the in-game API, the LCU API, Meraki's champion/item json data, eSports data, ... TFT?  If Cass's role is to represent all League of Legends data, we're falling far short of that. But we may not want Cass to support all of that functionality. For example, @sousa-andre has a great app for working with the LCU API. In theory, Cass could use Andre's library as a dependency to allow users to access their local League of Legends data, and maybe Cass should do that. It would be pretty slick. The in-game API is just a simple http endpoint that hits a localhost url, and it could be integrated along side the LCU API functionality. eSports data is a different matter, and @mrtolkien has recently put together a new JSON schema to represent League of Legends eSports games, along with python code to get ahold of that eSports data. We could integrate this functionality into Cass in a similar way, so that Cass would support all eSports games. That would be sweet! Finally, we've been wanting to integrate Meraki's champion/item json data into Cass as an alternative to DDragon's static data. And that brings up another piece of League data... should Cass support CDragon?  --  You get the idea.

These are all open questions, but the relevant one for this discussion is TFT.   I'm going to sideline that question for another minute and ask... Should Cass support all Riot games? Rob and I have decided that the answer to that question is without a doubt "No!" for reasons I'll describe in a few minutes.  So now I'll ask: Should Cass support TFT? i.e. is TFT part of League of Legends? In a way TFT _is_ a part of League of Legends: You play it through the same client, it has the same interface, and some of its data overlaps with League of Legends. All of that says "yes" TFT is part of League of Legends. But recently Riot has release [TFT mobile](https://play.google.com/store/apps/details?id=com.riotgames.league.teamfighttactics&hl=en_US) and put out a [press release](https://wildrift.leagueoflegends.com/en-us/) for [Wild Rift](https://play.google.com/store/apps/details?id=com.riotgames.league.wildrift&hl=en_US). In this instance, Riot has created two completely separate platforms for League of Legends and TFT. I think that TFT's tight integration with League of Legends is largely a consequence of Riot wanting to put the "s" on Riot Games, and get their first new game out the door to test it with their players. I think that Riot, and the community, sees League of Legends and TFT as two separate but related games. For example, while League of Legends and TFT have the same champions, their skills and stats are all largely unrelated.

You can see where this is going... So, should Cass support TFT? Our opinion is that no, Cass (and it's sister library Orianna) should not support TFT. TFT is its own game and therefore TFT deserves its own dedicated API library. This has more positives than it does negatives.

So why shouldn't Cass support TFT, beyond the fact that it is a "separate game"?

1. It limits the scope of Cass to a more manageable set of code. Cass is already massive, and while it's pretty stable, it is difficult if not impossible to wrap your head around the entirety of Cass. Even after such a major contribution as you have made, I doubt that you feel like you really understand everything about Cass. The truth is that I don't either. It's been three years since Rob and I wrote Cass and Ori, and I've forgotten the details. It requires such little maintenance that we are able to step away and let Cass do its thing without needing to constantly know everything there is to know about it. This is nice for us! Limiting the scope of Cass also makes Cass more understandable to the rest of the community. It's already massive, and to feel comfortable creating a PR takes a _ton_ of pre-code work (before you even start, you have to figure out what the heck is going on). All of this is to say that we want to keep Cass at a maintainable level, both for ourselves and for the community.

2. Allowing TFT to have it's own library, not maintained by Meraki, allows other people to participate in the Riot API community. One thing I have conflicting thoughts about is the infrastructure the developer community has built up around the Riot API. On one hand, I'm incredibly proud that we've built such awesome API libraries and other utilities for the API. I'd say that ~ 90% of the common use-cases for using the API are covered (or were a year ago). Five years ago, _everyone_ wanting to use the Riot API had to write their own rate limiter. It was such an absurd replication of effort by all of us. Now, every single API wrapper provides rate limiting for its users. API wrappers exist for every single common language. We also have LCU API libraries (Andre's and others). ... Role identification ... the list goes on. The developer community has created this incredible suite of tools for other developers to use, and it's awesome!  On the other side of the coin, because 90% of the stuff has been done, the developer community is pretty stagnant. It used to be the case where we were all rushing to create things, build cool tools, competing with each other and creating awesome products in the process. I miss that, _a lot_. In a way it's sad to me to see the developer community not creating products that _everyone_ can use (rather than creating websites, apps, etc. - which are also cool but in a very different way). ---- SO! _This_ is the opportunity to create something new. The Valorant API has sparked a resurgence of excitement around the Riot API, and if the Valorant API was open to everyone, the discord and the API community would be a whirlwind of fun. TFT doesn't have quite the same hype, but it has the potential to seize some of the same passion and excitement. I desperately want to see this. And I don't want to do it. The gurus and a few of the other members of the developer community make and maintain all of the tools for the API, and the truth is that many of us are starting to go in different directions in life than we were five+ years ago when we started. I want to see fresh faces, creating cool stuff, taking responsibility for supporting the community. A TFT library is a great place to do this. That said, I would be **thrilled** to see the TFT library emulate Cass and use the same underlying technology (e.g. Meraki's ghost objects, the data pipeline, etc.) because I'm super proud of Cass and Ori's interface and I think the rest of the API community is as well. - Okay I'll stop writing about this, but clearly you can tell how important this is to me.

3. Related to both (1) and (2), making TFT have it's own library reduces the load on Rob and me. One of the great things about providing great API libraries for the community is that people want to use and improve them. That can take a ton of time, and while we love doing it and discussing it (as you can see by the length of this comment!), it can also get to be a lot of work. I'd like to distribute that workload to other people in the community. (In fact, Rob and I were just talking about trying to find some people who will seriously help to maintain Cass and Ori.) Just the mental overhead of having to keep track of everything is starting to be a lot. But it's very very hard to find maintainers while still having some say in the direction of the tool. All of a sudden the decision making goes from a dictatorship to a democracy, and that makes it much harder to make decisions. Creating a completely separate TFT library sidesteps the maintenance/responsibility question altogether.

Those are the only three/four reasons I can think of right now. I think they are pretty powerful reasons.


Okay, so what now?  I would love to see you git-clone this PR, strip out everything non-League of Legends related, and push it to your own repository. Take it. Own it. Maintain it. Tell people about it. Give us a little tiny bit of credit if you want for using Cass as an inspiration and for using some of Meraki's tools behind the scenes :) But really, let this be yours. If you're willing to do it and you want to continue to be part of the developer community, I think this would be a great way to take part in the community, provide value, learn, help people, and _own_ something of your own that people use. It's a powerful thing. I think by taking what you already have and stripping out the League of Legends data/classes/information/etc., that you will be left with a stripped down version of this library that has some fantastic functionality, and you'll have full control over it and its future.

What do you think?

**Comment by paaksing** [comment]
I agree totally with Jason even before reading, but was waiting for his reaction about this topic.
TFT def is a different game and should have its own framework to work with, even the Riot teams prefers you to have 2 different production key for those 2 games (I know it can be merged, but why ?).

**Comment by Crimack** [comment]
Thanks for taking the time to write such a detailed comment, and for continuing to actively maintain these API wrappers. I appreciate that communicating well takes a lot more time and effort than you'd think, and it's a well thought out response.

I fully agree with all of your points. Honestly I just found myself with a spare hour or two and decided to try and be slightly productive, so I lifted an issue from a project I've enjoyed using in the past without giving it too much thought. It did take a good bit of time for me to understand how the whole library fit together (and I was making a reasonably simple change with another example that was almost the same). Adding another, very similar looking set of objects for a very similar API wouldn't have helped if I was coming to the code base for the first time.

Plus, this PR would confuse the Cass API quite a bit by introducing another object that's _almost_ the same but behaves differently and isn't interchangeable. Having two separate abstractions for what's really the same Riot account in the same code base is probably a bad idea. I think your reasoning for not merging is sound. 

I think having a separate TFT library sounds like a solid idea. I've got a little free time on my hands right now, so I'll try have a go at it over the next week or so (and probably using the same stack as Cass so I don't have to think _too_ hard), assuming life doesn't get in the way. It seems like AP mages is the standard naming convention, so I'm feeling Lissandra 😊

**Comment by paaksing** [comment]
@Crimack Currenty cassiopeia usage is extended to Python Web Framework Django too, if you get the chance to finish the library, I can extend it the same way as cassiopeia does. Happy Coding ~

**Comment by Crimack** [comment]
Gonna go ahead and close this PR, rather than leaving it lingering around in the wild

I managed to make a bit of progress on a TFT library, which is half alive at https://github.com/Crimack/lissandra. Ripping out unused code took a bit longer than expected and life stuff happened, it supports the TFT summoner and league APIs right now. I'll hopefully get around to adding match/match history stuff soonish :)

**Pull request closed without merging**